tsdb/shard.go

```diff
@@ -1081,13 +1153,19 @@ func (a Shards) Less(i, j int) bool { return a[i].id < a[j].id }
 // Swap implements sort.Interface.
 func (a Shards) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
 
+// MeasurementsByRegex returns the unique set of measurements matching the
+// provided regex, for all the shards.
 func (a Shards) MeasurementsByRegex(re *regexp.Regexp) []string {
 	m := make(map[string]struct{})
 	for _, sh := range a {
-		names := sh.MeasurementsByRegex(re)
-		for _, name := range names {
-			m[name] = struct{}{}
+		sh.mu.RLock()
+		if err := sh.ready(); err == nil {
+			names := sh.measurementsByRegex(re)
```

Comment:
Can you have this call `MeasurementsByRegex` instead of accessing the `sh.mu` directly outside of `Shard`?